mesonbuild/interpreter.py

```diff
@@ -3616,8 +3662,32 @@ def dependency_fallback(self, name, display_name, kwargs):
             'required': required,
         }
         self.do_subproject(dirname, 'meson', sp_kwargs)
+        self.override_dependencies_from_wrap(dirname, kwargs)
         return self.get_subproject_dep(name, display_name, dirname, varname, kwargs)
 
+    def override_dependencies_from_wrap(self, dirname, kwargs):
+        # Make sure to override every dependencies this wrap provides in the
+        # case the subproject did not call meson.override_dependency().
+        # This ensures that if we fallbacked for one dependency, any other
+        # dependency provided by this subproject won't be picked from system.
+        # Also verify consistency in the case the subproject overridden the
+        # dependency with a different object than the variable from wrap file.
+        for_machine = self.machine_from_native_kwarg(kwargs)
```

Comment:
I think for what machines something is overridden by a subproject is not related to what the using program wishes to obtain.

Actually i didn't think about cross compiling at all, but this really forces us to find a solution that is actually correct. Sorry for missing this the whole time.

This might work when everything matches very well and no edge cases show up, but i don't think just hoping for it is ok. Maybe we need provide_xxx sections to do this properly. Or maybe there is a neat simple solution out there.

Reply:
I think this opens a whole new source of issues that are not trivial at all. IMHO we could ignore that for now and think about a solution in a separate PR, is that fine with you?

Note that it is already causing issues, nothing new here:
```
d = dependency('foo', fallback : ['foo', 'dep'], native : true)
e = executable('app', 'app.c', native : true, dependencies : d)
```
When cross compiling that produce that error:
```
meson.build:4:0: ERROR: Tried to mix libraries for machines MachineChoice.BUILD and MachineChoice.HOST in target 'app' This is not possible in a cross build.
```

Reply:
I seem to remember we discussed this in irc and came to the agreement to use the machine for which the dependency was compiled. See https://github.com/mesonbuild/meson/pull/6902#issuecomment-633044173 for my notes.

With subprojects properly overriding the dependency, things should work because override_dependency does the plumbing machine specific. I think here we should do the plumbing based on the machine the dependency was created for. As discussed on irc this means we might need to add a bit of info to the objects we can use here. But it does not make any sense to plumb dependencies cross machine. What we want here is a proper error message.

Reply:
Hmm... The problem is the variable is likely an InternalDependency object, created with declare_dependency(), and there is no for_machine associated with that. Also get_provided_deps() does not tell for which machine each are so we don't know in which self.build.dependency_overrides dict we need to check for cached_dep, we would have to extend wrap syntax for that... Getting more complicated, I would keep that out of scope for now tbh. I'm starting to think that override_dependencies_from_wrap() is a can of worms, maybe I should just delete that function completely.

Reply:
I don't think we need to complicate the .wrap information just now.
What i was thinking was that we should extend the dependency objects to store for which machine their parts are compiled. After executing a subproject meson would go through all variables from the wrap file and determine the machine they are for and then use that to override the named dependency in that machine. Here we can check that each named library gets an override. If in the machine the dependency was requested any of the dependency names are not overriden meson should mark them as not-found.

That keeps the information in the .wrap simple and if we can somehow add a message to those not found deps it should lead to a sensible UX as well. Cross aware subprojects can just use override_dependency for both machines where that is useful and should be able to make the important cases work for consumers.

Does that make sense?

Reply:
I think it's getting complicated with has corner cases, I would like to get this PR merged first and think a bit more about this afterward. I update the PR to drop that function completely. 

A few points:
- This whole function is only covering a really corner case where a subproject provide >=2 dependencies and we fallback for one of them but the other is available on the system and we want to force fallback for that one too. In practice that's pretty rare to happen, and even when it does the proper fix is already for the subproject to call meson.override_dependency().
- When the wrap file does not have the varname, which is what we are aiming for, we have no way to guess for which machine it is. So ideally projects should still override the dependencies they optionally don't provide with not-found otherwise they are forcing main project to know the varname. Also most of the time in that case the variable won't even exist anyway... Again, this is really corner case.
- Would require a bit of work to ensure that every dependency object has self.for_machine. For InternalDependency it means guessing it from BuildTarget in its link_with list, if any.